build.gradle

```diff
@@ -252,6 +251,107 @@ tasks.register('printProjects') {
     })
 }
 
+cyclonedxBom {
+    includeConfigs = ["runtimeClasspath"]
+}
+
+tasks.register('generateDistLicense') {
+    dependsOn('cyclonedxBom') // Task from 'org.cyclonedx.bom' plugin
+    doLast {
+        // Inputs
+        def bomFile = file("$buildDir/reports/bom.json")
+        def bom = new JsonSlurper().parseText(bomFile.text)
+        def projectLicenseText = file('LICENSE').text
+
+        // Outputs
+        def distLicenseFile = file('tools/dist-license/LICENSE')
+        def licensesDir = file('tools/dist-license/licenses/java/')
+        if (licensesDir.exists()) {
+            licensesDir.eachFile { it.delete() }
+        } else {
+            licensesDir.mkdirs()
+        }
+
+        List<Map<String, String>> thirdPartyArtifacts = new ArrayList<Map<String, String>>()
+        // Parse BOM
+        bom.components.each { component ->
+            // Exclude project modules
+            if (!component.group.startsWith('org.apache.eventmesh')) {
+                component.licenses.each { artifactLicense ->
+                    if (artifactLicense.license != null) {
+                        Map<String, String> artifact = new HashMap<String, String>()
+                        artifact.put("name", component.name)
+                        artifact.put("version", component.version)
+                        if (artifactLicense.license.id != null) {
+                            artifact.put("license", artifactLicense.license.id)
+                            if (artifactLicense.license.text != null) {
+                                artifact.put("licenseContent", new String(artifactLicense.license.text.content.decodeBase64()))
+                            }
+                        } else {
+                            artifact.put("license", artifactLicense.license.name)
+                            artifact.put("licenseContent", artifactLicense.license.url)
+                        }
+                        thirdPartyArtifacts.add(artifact)
+                    }
+                }
+            }
+        }
+        thirdPartyArtifacts.sort { a, b ->
+            def nameComparison = a.name <=> b.name
+            if (nameComparison == 0) {
+                return a.version <=> b.version
+            } else {
+                return nameComparison
+            }
+        }
+
+        def distLicenseText = projectLicenseText + "\n=======================================================================\n" +
+                "This distribution contains the following third-party artifacts:\n\n"
+        thirdPartyArtifacts.each { artifact ->
+            // Write licenses
+            def artifactLicenseFilename = artifact.license.replaceAll("/", "-") + ".txt"
+            def artifactLicenseFile = new File(licensesDir, artifactLicenseFilename)
```

Comment:
I think that it would be safer **not** to generate the contents of `tools/dist-license/licenses/java` each time, but to have a directory of approved licenses (let's say `tools/approved-licenses`), where each license is in a file named:
```
tools/approved-licenses/<SPDX identifier>.txt
```
where `<SPDX identifier>` is a [SPDX identifier of the license](https://spdx.org/licenses/).

This task will then:

1. For each artifact determine the SPDX identifier of the license,
2. If the identifier is known and a corresponding `tools/approved-licenses/<SPDX identifier>.txt` exist, copy the file to the `tools/dist-license/licenses/java` directory,
3. Otherwise fail and ask the user to fix the error.

In order to determine the SPDX identifier of the license of each artifact you can:
1. Check if `artifactLicense.license.id` exist,
2. If it doesn't exist consult a manually written map between license URLs and license IDs. This map could be for example in an external JSON file. It only needs to contain the exceptions, e.g. the URL `http://antlr.org/license.html` should be mapped to `BSD-3-Clause`,
3. If the artifact does not have an URL for the license, consult a map that associates `${artifact.group}:${artifact:name}` and the license ID.

Reply:
Your suggestion is actually a change from "license blacklisting" to "license whitelisting". But ASF approves mainstream licenses **and their variants**, and according to the [SPDX list](https://spdx.org/licenses/), it's extremely difficult to maintain a complete list of these variants, not to mention that there are a lot of legal artifacts that don't comply with SPDX, so I think a blacklist is more appropriate for an open source project.

Reply:
You don't need the complete list, you only need a bunch of license, which are already in the `tools/dist-license/licenses/java` folder of this PR. And you also have many duplicates: e.g. the wording of https://www.bouncycastle.org/licence.html is exactly the same as the MIT license.

When it comes to artifacts that don't comply with SPDX, IMHO they can not be included in an Open Source distribution until a lawyer proof-reads the license. In this case you might as well wait until the license gets a SPDX identifier and is vetted by the ASF.

Reply:
>When it comes to artifacts that don't comply with SPDX, in my humble opinion, they should not be included in an Open Source distribution until a lawyer has reviewed the license.

The majority of artifacts that don't fully comply with SPDX are usually compatible. For example, the "ASL" license of `jna` and the "AL 2.0" license of `amqp-client` are essentially the same as Apache-2.0. However, the Amazon Software License, which is also an ASL, is not compatible. Therefore, we must treat artifacts differently based on whether they use SPDX identifiers or custom names (which has been addressed in this pull request by handling `license.id` and `license.name` separately).

>These artifacts can be found in the `tools/dist-license/licenses/java` folder of this pull request.

The `generateDistLicense` Gradle task does not guarantee that the Release Manager can completely ignore license checks. It is the responsibility of the RM to ensure the compatibility of all licenses. The purpose of submitting this PR is to reduce the workload of the RM and ensure that the licenses of all project dependencies are included in the distribution and not overlooked.

If the logic of the `generateDistLicense` task is changed to a whitelist mode, it would mean that every time we release, we would have to manually approve all dependencies from the past six months by creating new files and pasting the license contents, and remove licenses that are no longer included in this version.

If we maintain the existing logic of this PR, the RM would **only need to check for newly added files in Git** and confirm compatibility with the added licenses. Compared to the former approach, I believe the latter is easier and equally secure. Therefore, I think maintaining the existing generation logic of this PR is sufficient.